File: session_rust/src/color_test.rs
#[cfg(test)]
mod tests {
    use crate::Color;

    #[test]
    fn test_color_constructor() {
        let mut red = Color::new(255, 0, 0, 255);
        red.name = "red".to_string();
        assert_eq!(red.name, "red");
        assert!(!red.guid.to_string().is_empty());
        assert_eq!(red.r, 255);
        assert_eq!(red.g, 0);
        assert_eq!(red.b, 0);
        assert_eq!(red.a, 255);
    }

    #[test]
    fn test_color_equality() {
        let c1 = Color::new(0, 100, 50, 200);
        let c2 = Color::new(0, 100, 50, 200);
        // Colors have different GUIDs, so they won't be equal by default
        // We'll compare the RGBA values instead
        assert_eq!(c1.r, c2.r);
        assert_eq!(c1.g, c2.g);
        assert_eq!(c1.b, c2.b);
        assert_eq!(c1.a, c2.a);

        let c3 = Color::new(0, 100, 50, 200);
        let c4 = Color::new(1, 100, 50, 200);
        assert_ne!(c3.r, c4.r);
    }

    #[test]
    fn test_color_to_json_data() {
        let mut color = Color::new(128, 64, 192, 255);
        color.name = "purple".to_string();

        let json_string = color.to_json_data().unwrap();
        let data: serde_json::Value = serde_json::from_str(&json_string).unwrap();

        assert_eq!(data["type"], "Color");
        assert_eq!(data["name"], "purple");
        assert_eq!(data["r"], 128);
        assert_eq!(data["g"], 64);
        assert_eq!(data["b"], 192);
        assert_eq!(data["a"], 255);
        assert!(data["guid"].is_string());
    }

    #[test]
    fn test_color_from_json_data() {
        let mut original_color = Color::new(200, 150, 100, 255);
        original_color.name = "bronze".to_string();

        let json_string = original_color.to_json_data().unwrap();
        let restored_color = Color::from_json_data(&json_string).unwrap();

        assert_eq!(restored_color.r, 200);
        assert_eq!(restored_color.g, 150);
        assert_eq!(restored_color.b, 100);
        assert_eq!(restored_color.a, 255);
        assert_eq!(restored_color.name, "bronze");
        assert_eq!(restored_color.guid, original_color.guid);
    }

    #[test]
    fn test_color_to_json_from_json() {
        let mut original = Color::new(255, 128, 64, 255);
        original.name = "sunset_orange".to_string();
        let filename = "test_color.json";

        original.to_json(filename).unwrap();
        let loaded = Color::from_json(filename).unwrap();

        assert_eq!(loaded.r, original.r);
        assert_eq!(loaded.g, original.g);
        assert_eq!(loaded.b, original.b);
        assert_eq!(loaded.a, original.a);
        assert_eq!(loaded.name, original.name);
        assert_eq!(loaded.guid, original.guid);
    }

    #[test]
    fn test_color_white() {
        let white = Color::white();
        assert_eq!(white.name, "white");
        assert_eq!(white.r, 255);
        assert_eq!(white.g, 255);
        assert_eq!(white.b, 255);
        assert_eq!(white.a, 255);
    }

    #[test]
    fn test_color_black() {
        let black = Color::black();
        assert_eq!(black.name, "black");
        assert_eq!(black.r, 0);
        assert_eq!(black.g, 0);
        assert_eq!(black.b, 0);
        assert_eq!(black.a, 255);
    }

    #[test]
    fn test_color_to_float_array() {
        let color = Color::new(255, 128, 64, 255);
        let float_array = color.to_float_array();
        assert_eq!(float_array, [1.0, 0.5019608, 0.2509804, 1.0]);
    }

    #[test]
    fn test_color_from_float() {
        let color = Color::from_float(1.0, 0.5, 0.25, 1.0);
        assert_eq!(color.r, 255);
        assert_eq!(color.g, 128); // 0.5 * 255 = 127.5, rounded to 128 in Rust
        assert_eq!(color.b, 64); // 0.25 * 255 = 63.75, rounded to 64
        assert_eq!(color.a, 255);
    }
}

File: session_py/src/session_py/color_test.py
from .color import Color


def test_color_constructor():
    """Test Color constructor."""
    red = Color(255, 0, 0, 255, "red")
    assert red.name == "red"
    assert red.guid != ""
    assert red.r == 255
    assert red.g == 0
    assert red.b == 0
    assert red.a == 255


def test_color_equality():
    """Test Color equality."""
    c1 = Color(0, 100, 50, 200)
    c2 = Color(0, 100, 50, 200)
    assert c1 == c2
    assert not (c1 != c2)

    c3 = Color(0, 100, 50, 200)
    c4 = Color(1, 100, 50, 200)
    assert not (c3 == c4)
    assert c3 != c4


def test_color_to_json_data():
    """Test Color to_json_data method."""
    color = Color(128, 64, 192, 255, "purple")
    data = color.to_json_data()
    assert data["type"] == "Color"
    assert data["name"] == "purple"
    assert data["r"] == 128
    assert data["g"] == 64
    assert data["b"] == 192
    assert data["a"] == 255
    assert "guid" in data


def test_color_from_json_data():
    """Test Color from_json_data method."""
    original_color = Color(200, 150, 100, 255, "bronze")
    data = original_color.to_json_data()
    restored_color = Color.from_json_data(data)
    assert restored_color.r == 200
    assert restored_color.g == 150
    assert restored_color.b == 100
    assert restored_color.a == 255
    assert restored_color.name == "bronze"
    assert restored_color.guid == original_color.guid


def test_color_to_json_from_json():
    """Test Color file I/O with to_json and from_json."""
    original = Color(255, 128, 64, 255, "sunset_orange")
    filename = "test_color.json"

    original.to_json(filename)
    loaded = Color.from_json(filename)

    assert loaded.r == original.r
    assert loaded.g == original.g
    assert loaded.b == original.b
    assert loaded.a == original.a
    assert loaded.name == original.name
    assert loaded.guid == original.guid


def test_color_white():
    """Test Color.white() class method."""
    white = Color.white()
    assert white.name == "white"
    assert white.r == 255
    assert white.g == 255
    assert white.b == 255
    assert white.a == 255


def test_color_black():
    """Test Color.black() class method."""
    black = Color.black()
    assert black.name == "black"
    assert black.r == 0
    assert black.g == 0
    assert black.b == 0
    assert black.a == 255


def test_color_to_float_array():
    """Test Color to_float_array method."""
    color = Color(255, 128, 64, 255)
    float_array = color.to_float_array()
    assert float_array == [1.0, 0.5019607843137255, 0.25098039215686274, 1.0]


def test_color_from_float():
    """Test Color.from_float() class method."""
    color = Color.from_float(1.0, 0.5, 0.25, 1.0)
    assert color.r == 255
    assert color.g == 127  # 0.5 * 255 = 127.5, rounded to 127
    assert color.b == 63  # 0.25 * 255 = 63.75, rounded to 63
    assert color.a == 255


def test_color_red():
    """Test Color.red() class method."""
    red = Color.red()
    assert red.name == "red"
    assert red.r == 255
    assert red.g == 0
    assert red.b == 0
    assert red.a == 255


def test_color_green():
    """Test Color.green() class method."""
    green = Color.green()
    assert green.name == "green"
    assert green.r == 0
    assert green.g == 255
    assert green.b == 0
    assert green.a == 255


def test_color_blue():
    """Test Color.blue() class method."""
    blue = Color.blue()
    assert blue.name == "blue"
    assert blue.r == 0
    assert blue.g == 0
    assert blue.b == 255
    assert blue.a == 255


def test_color_grey():
    """Test Color.grey() class method."""
    grey = Color.grey()
    assert grey.name == "grey"
    assert grey.r == 128
    assert grey.g == 128
    assert grey.b == 128
    assert grey.a == 255
